File: service/mrdocument/costs.py
"""
Cost tracking module for Anthropic API usage.

Tracks Anthropic token usage per model and calculates costs based on pricing config.
Uses an in-memory queue with periodic flush to avoid file locking issues.
"""

import asyncio
import json
import logging
import os
from dataclasses import dataclass
from datetime import date
from pathlib import Path
from queue import Queue, Empty
from threading import Thread
from typing import Optional

import yaml

logger = logging.getLogger(__name__)


@dataclass
class AnthropicUsage:
    """Record of Anthropic API usage."""
    model: str
    input_tokens: int
    output_tokens: int
    user_dir: Path
    count_document: bool = False  # If True, increment document counter


class CostTracker:
    """
    Tracks API costs with in-memory queue and periodic flush.
    
    Each service (mrdocument, stt) should have its own CostTracker instance
    writing to a separate file to avoid cross-container conflicts.
    """
    
    def __init__(
        self,
        filename: str = "mrdocument_costs.json",
        flush_interval: float = 30.0,
        pricing_config_path: Optional[Path] = None,
    ):
        """
        Initialize cost tracker.
        
        Args:
            filename: Name of the costs file (created in each user's dir)
            flush_interval: Seconds between flushes to disk
            pricing_config_path: Path to pricing.yaml config
        """
        self.filename = filename
        self.flush_interval = flush_interval
        self.queue: Queue = Queue()
        self._running = False
        self._flush_thread: Optional[Thread] = None
        
        # Load pricing config
        if pricing_config_path is None:
            pricing_config_path = Path(__file__).parent / "pricing.yaml"
        self.pricing = self._load_pricing(pricing_config_path)
        
        logger.info("CostTracker initialized: file=%s, flush_interval=%.1fs", filename, flush_interval)
    
    def _load_pricing(self, path: Path) -> dict:
        """Load pricing configuration from YAML."""
        if not path.exists():
            logger.warning("Pricing config not found at %s, using empty pricing", path)
            return {"anthropic": {}}
        
        try:
            with open(path) as f:
                return yaml.safe_load(f) or {"anthropic": {}}
        except Exception as e:
            logger.error("Failed to load pricing config: %s", e)
            return {"anthropic": {}}
    
    def start(self) -> None:
        """Start the background flush thread."""
        if self._running:
            return
        
        self._running = True
        self._flush_thread = Thread(target=self._flush_loop, daemon=True)
        self._flush_thread.start()
        logger.info("CostTracker flush thread started")
    
    def stop(self) -> None:
        """Stop the flush thread and perform final flush."""
        if not self._running:
            return
        
        self._running = False
        if self._flush_thread:
            self._flush_thread.join(timeout=5.0)
        
        # Final flush
        self._flush_all()
        logger.info("CostTracker stopped")
    
    def record_anthropic(
        self,
        model: str,
        input_tokens: int,
        output_tokens: int,
        user_dir: Path,
        count_document: bool = False,
    ) -> None:
        """
        Record Anthropic API usage.
        
        Args:
            model: Model name (e.g., "claude-sonnet-4-5")
            input_tokens: Number of input tokens
            output_tokens: Number of output tokens
            user_dir: User's directory where costs.json will be stored
            count_document: If True, increment the document counter
        """
        usage = AnthropicUsage(
            model=model,
            input_tokens=input_tokens,
            output_tokens=output_tokens,
            user_dir=user_dir,
            count_document=count_document,
        )
        self.queue.put(usage)
        logger.debug(
            "Queued Anthropic usage: model=%s, in=%d, out=%d, user=%s, count_doc=%s",
            model, input_tokens, output_tokens, user_dir.name, count_document
        )
    
    def _flush_loop(self) -> None:
        """Background loop that periodically flushes the queue."""
        while self._running:
            try:
                asyncio.run(asyncio.sleep(self.flush_interval))
            except Exception:
                pass
            
            if self._running:
                self._flush_all()
    
    def _flush_all(self) -> None:
        """Flush all queued usage records to disk."""
        # Collect all pending records grouped by user_dir
        records_by_user: dict[Path, list] = {}
        
        while True:
            try:
                usage = self.queue.get_nowait()
                user_dir = usage.user_dir
                if user_dir not in records_by_user:
                    records_by_user[user_dir] = []
                records_by_user[user_dir].append(usage)
            except Empty:
                break
        
        if not records_by_user:
            return
        
        # Write to each user's costs file
        today = date.today().isoformat()
        
        for user_dir, records in records_by_user.items():
            try:
                self._write_user_costs(user_dir, today, records)
            except Exception as e:
                logger.error("Failed to write costs for %s: %s", user_dir.name, e)
    
    def _write_user_costs(self, user_dir: Path, today: str, records: list) -> None:
        """Write usage records to a user's costs file."""
        # Extract username from user_dir (e.g., /sync/alice/mrdocument -> alice)
        # Go up to find the user folder under /sync
        username = None
        for parent in [user_dir] + list(user_dir.parents):
            if parent.parent.name == "sync" or str(parent.parent) == "/sync":
                username = parent.name
                break
        
        if not username:
            # Fallback: use the parent directory name
            username = user_dir.parent.name if user_dir.parent.name != "sync" else user_dir.name
        
        # Write to /costs/{username}/
        costs_dir = Path("/costs") / username
        costs_dir.mkdir(parents=True, exist_ok=True)
        costs_path = costs_dir / self.filename
        
        # Load existing data
        if costs_path.exists():
            try:
                with open(costs_path) as f:
                    data = json.load(f)
            except (json.JSONDecodeError, Exception):
                data = {}
        else:
            data = {}
        
        # Ensure today's entry exists
        if today not in data:
            data[today] = {"anthropic": {}}

        day_data = data[today]

        # Aggregate records
        for usage in records:
            if isinstance(usage, AnthropicUsage):
                if "anthropic" not in day_data:
                    day_data["anthropic"] = {}
                
                # Get or create model-specific entry
                model_key = usage.model
                if model_key not in day_data["anthropic"]:
                    day_data["anthropic"][model_key] = {
                        "input_tokens": 0, "output_tokens": 0, "cost": 0.0, "documents": 0
                    }
                model_data = day_data["anthropic"][model_key]
                
                model_data["input_tokens"] += usage.input_tokens
                model_data["output_tokens"] += usage.output_tokens
                if usage.count_document:
                    model_data["documents"] += 1
                
                # Calculate cost
                cost = self._calculate_anthropic_cost(
                    usage.model, usage.input_tokens, usage.output_tokens
                )
                model_data["cost"] += cost
        
        # Round costs to avoid floating point artifacts
        if "anthropic" in day_data:
            for model_data in day_data["anthropic"].values():
                model_data["cost"] = round(model_data["cost"], 6)
        
        # Recalculate totals across all days
        total: dict = {"anthropic": {}}
        for date_key, date_data in data.items():
            if date_key == "total":
                continue
            
            # Aggregate anthropic per model
            if "anthropic" in date_data:
                for model_name, model_stats in date_data["anthropic"].items():
                    if model_name not in total["anthropic"]:
                        total["anthropic"][model_name] = {
                            "input_tokens": 0, "output_tokens": 0, "cost": 0.0, "documents": 0
                        }
                    total["anthropic"][model_name]["input_tokens"] += model_stats.get("input_tokens", 0)
                    total["anthropic"][model_name]["output_tokens"] += model_stats.get("output_tokens", 0)
                    total["anthropic"][model_name]["cost"] += model_stats.get("cost", 0.0)
                    total["anthropic"][model_name]["documents"] += model_stats.get("documents", 0)
        
        # Calculate cost per document and round costs
        for model_name, model_stats in total["anthropic"].items():
            model_stats["cost"] = round(model_stats["cost"], 6)
            if model_stats["documents"] > 0:
                model_stats["cost_per_document"] = round(model_stats["cost"] / model_stats["documents"], 6)
            else:
                model_stats["cost_per_document"] = 0.0
        
        data["total"] = total
        
        # Write atomically via temp file
        temp_path = costs_path.with_suffix(".tmp")
        with open(temp_path, "w") as f:
            json.dump(data, f, indent=2)
        temp_path.rename(costs_path)
        
        logger.debug("Flushed %d records to %s", len(records), costs_path)
    
    def _calculate_anthropic_cost(self, model: str, input_tokens: int, output_tokens: int) -> float:
        """Calculate cost for Anthropic usage."""
        pricing = self.pricing.get("anthropic", {}).get(model)
        if not pricing:
            logger.warning("No pricing found for Anthropic model: %s", model)
            return 0.0
        
        input_cost = (input_tokens / 1_000_000) * pricing.get("input_per_1m", 0)
        output_cost = (output_tokens / 1_000_000) * pricing.get("output_per_1m", 0)
        return input_cost + output_cost


# Global instance for MrDocument
_cost_tracker: Optional[CostTracker] = None


def get_cost_tracker() -> CostTracker:
    """Get the global cost tracker instance, creating if needed."""
    global _cost_tracker
    if _cost_tracker is None:
        _cost_tracker = CostTracker()
        _cost_tracker.start()
    return _cost_tracker


def shutdown_cost_tracker() -> None:
    """Shutdown the global cost tracker."""
    global _cost_tracker
    if _cost_tracker is not None:
        _cost_tracker.stop()
        _cost_tracker = None

File: service-rs/src/costs.rs
use serde_json::{json, Value};
use std::collections::HashMap;
use std::io::Write;
use std::path::{Path, PathBuf};
use std::sync::Mutex;
use tracing::{debug, error, info, warn};

use crate::config::{load_pricing, PricingConfig};

struct CostRecord {
    model: String,
    input_tokens: u64,
    output_tokens: u64,
    user_dir: PathBuf,
    count_document: bool,
}

pub struct CostTracker {
    filename: String,
    pricing: PricingConfig,
    queue: Mutex<Vec<CostRecord>>,
    api_log: Mutex<Option<std::fs::File>>,
}

impl CostTracker {
    pub fn new() -> Self {
        let pricing = load_pricing(None);
        let api_log_file = std::env::var("LOG_DIR")
            .ok()
            .map(|dir| {
                let path = PathBuf::from(dir).join("api_calls.log");
                std::fs::OpenOptions::new()
                    .create(true)
                    .append(true)
                    .open(&path)
                    .map_err(|e| warn!("Failed to open API call log {:?}: {}", path, e))
                    .ok()
            })
            .flatten();
        info!("CostTracker initialized (api_log={})", api_log_file.is_some());
        Self {
            filename: "mrdocument_costs.json".to_string(),
            pricing,
            queue: Mutex::new(Vec::new()),
            api_log: Mutex::new(api_log_file),
        }
    }

    pub fn record_anthropic(
        &self,
        model: &str,
        input_tokens: u64,
        output_tokens: u64,
        user_dir: &Path,
        count_document: bool,
    ) {
        if let Ok(mut q) = self.queue.lock() {
            q.push(CostRecord {
                model: model.to_string(),
                input_tokens,
                output_tokens,
                user_dir: user_dir.to_path_buf(),
                count_document,
            });
        }
        debug!(
            "Recorded Anthropic usage: model={}, in={}, out={}, count_doc={}",
            model, input_tokens, output_tokens, count_document
        );
        self.flush();
    }

    /// Log an API call with truncated request body to the api_calls.log file.
    pub fn log_api_call(
        &self,
        operation: &str,
        model: &str,
        filename: Option<&str>,
        request_body: &serde_json::Value,
        input_tokens: u64,
        output_tokens: u64,
        user_dir: &Path,
    ) {
        let cost = self.calculate_cost(model, input_tokens, output_tokens);

        // Truncate each string field in the request body to 10 chars
        let truncated = truncate_json_fields(request_body, 10);

        let timestamp = chrono::Utc::now().format("%Y-%m-%dT%H:%M:%S%.3fZ");
        let fname = filename.unwrap_or("-");
        let username = extract_username(user_dir);

        let line = format!(
            "{} | {} | {} | {} | {} | in={} out={} | cost=${:.6} | body={}\n",
            timestamp,
            username,
            operation,
            model,
            fname,
            input_tokens,
            output_tokens,
            cost,
            truncated,
        );

        if let Ok(mut guard) = self.api_log.lock() {
            if let Some(ref mut file) = *guard {
                let _ = file.write_all(line.as_bytes());
                let _ = file.flush();
            }
        }
    }

    pub fn flush(&self) {
        let records = {
            let mut q = match self.queue.lock() {
                Ok(q) => q,
                Err(_) => return,
            };
            std::mem::take(&mut *q)
        };

        if records.is_empty() {
            return;
        }

        let mut by_user: HashMap<PathBuf, Vec<&CostRecord>> = HashMap::new();
        for record in &records {
            by_user
                .entry(record.user_dir.clone())
                .or_default()
                .push(record);
        }

        let today = chrono::Local::now().format("%Y-%m-%d").to_string();
        for (user_dir, user_records) in by_user {
            if let Err(e) = self.write_user_costs(&user_dir, &today, &user_records) {
                error!("Failed to write costs for {:?}: {}", user_dir, e);
            }
        }
    }

    fn write_user_costs(
        &self,
        user_dir: &Path,
        today: &str,
        records: &[&CostRecord],
    ) -> Result<(), Box<dyn std::error::Error>> {
        // Find username
        let mut username = None;
        let mut current = Some(user_dir);
        while let Some(path) = current {
            if let Some(parent) = path.parent() {
                if parent.file_name().map(|n| n == "sync").unwrap_or(false)
                    || parent == Path::new("/sync")
                {
                    username = path.file_name().and_then(|n| n.to_str()).map(|s| s.to_string());
                    break;
                }
            }
            current = path.parent();
        }
        let username = username.unwrap_or_else(|| {
            user_dir
                .parent()
                .and_then(|p| p.file_name())
                .and_then(|n| n.to_str())
                .unwrap_or("unknown")
                .to_string()
        });

        let costs_dir = PathBuf::from("/costs").join(&username);
        std::fs::create_dir_all(&costs_dir)?;
        let costs_path = costs_dir.join(&self.filename);

        let mut data: Value = if costs_path.exists() {
            let content = std::fs::read_to_string(&costs_path)?;
            serde_json::from_str(&content).unwrap_or(json!({}))
        } else {
            json!({})
        };

        // Ensure today's entry
        if data.get(today).is_none() {
            data[today] = json!({"anthropic": {}});
        }

        // Aggregate
        for record in records {
            let day = &mut data[today];
            if day.get("anthropic").is_none() {
                day["anthropic"] = json!({});
            }
            let model_key = &record.model;
            if day["anthropic"].get(model_key).is_none() {
                day["anthropic"][model_key] = json!({
                    "input_tokens": 0,
                    "output_tokens": 0,
                    "cost": 0.0,
                    "documents": 0,
                });
            }
            let model_data = &mut day["anthropic"][model_key];
            let it = model_data["input_tokens"].as_u64().unwrap_or(0) + record.input_tokens;
            let ot = model_data["output_tokens"].as_u64().unwrap_or(0) + record.output_tokens;
            model_data["input_tokens"] = json!(it);
            model_data["output_tokens"] = json!(ot);
            if record.count_document {
                let docs = model_data["documents"].as_u64().unwrap_or(0) + 1;
                model_data["documents"] = json!(docs);
            }
            let cost = self.calculate_cost(&record.model, record.input_tokens, record.output_tokens);
            let prev_cost = model_data["cost"].as_f64().unwrap_or(0.0);
            model_data["cost"] = json!(((prev_cost + cost) * 1_000_000.0).round() / 1_000_000.0);
        }

        // Recalculate totals
        let mut total_anthropic: HashMap<String, Value> = HashMap::new();
        if let Some(obj) = data.as_object() {
            for (key, day_data) in obj {
                if key == "total" {
                    continue;
                }
                if let Some(anth) = day_data.get("anthropic").and_then(|v| v.as_object()) {
                    for (model_name, model_stats) in anth {
                        let entry = total_anthropic.entry(model_name.clone()).or_insert_with(|| {
                            json!({"input_tokens": 0, "output_tokens": 0, "cost": 0.0, "documents": 0})
                        });
                        let it = entry["input_tokens"].as_u64().unwrap_or(0)
                            + model_stats.get("input_tokens").and_then(|v| v.as_u64()).unwrap_or(0);
                        let ot = entry["output_tokens"].as_u64().unwrap_or(0)
                            + model_stats.get("output_tokens").and_then(|v| v.as_u64()).unwrap_or(0);
                        let cost = entry["cost"].as_f64().unwrap_or(0.0)
                            + model_stats.get("cost").and_then(|v| v.as_f64()).unwrap_or(0.0);
                        let docs = entry["documents"].as_u64().unwrap_or(0)
                            + model_stats.get("documents").and_then(|v| v.as_u64()).unwrap_or(0);
                        entry["input_tokens"] = json!(it);
                        entry["output_tokens"] = json!(ot);
                        entry["cost"] = json!((cost * 1_000_000.0).round() / 1_000_000.0);
                        entry["documents"] = json!(docs);
                        if docs > 0 {
                            entry["cost_per_document"] =
                                json!(((cost / docs as f64) * 1_000_000.0).round() / 1_000_000.0);
                        }
                    }
                }
            }
        }
        data["total"] = json!({"anthropic": total_anthropic});

        // Atomic write
        let tmp_path = costs_path.with_extension("tmp");
        std::fs::write(&tmp_path, serde_json::to_string_pretty(&data)?)?;
        std::fs::rename(&tmp_path, &costs_path)?;
        debug!("Flushed {} records to {:?}", records.len(), costs_path);

        Ok(())
    }

    fn calculate_cost(&self, model: &str, input_tokens: u64, output_tokens: u64) -> f64 {
        if let Some(pricing) = self.pricing.anthropic.get(model) {
            let input_cost = (input_tokens as f64 / 1_000_000.0) * pricing.input_per_1m;
            let output_cost = (output_tokens as f64 / 1_000_000.0) * pricing.output_per_1m;
            input_cost + output_cost
        } else {
            warn!("No pricing found for model: {}", model);
            0.0
        }
    }
}

/// Extract username from a user_dir path (looks for `/sync/{username}` pattern).
fn extract_username(user_dir: &Path) -> String {
    let mut current = Some(user_dir);
    while let Some(path) = current {
        if let Some(parent) = path.parent() {
            if parent.file_name().map(|n| n == "sync").unwrap_or(false)
                || parent == Path::new("/sync")
            {
                return path
                    .file_name()
                    .and_then(|n| n.to_str())
                    .unwrap_or("unknown")
                    .to_string();
            }
        }
        current = path.parent();
    }
    user_dir
        .file_name()
        .and_then(|n| n.to_str())
        .unwrap_or("unknown")
        .to_string()
}

/// Truncate string values in a JSON value to `max_len` chars for logging.
fn truncate_json_fields(value: &serde_json::Value, max_len: usize) -> String {
    match value {
        Value::Object(map) => {
            let entries: Vec<String> = map
                .iter()
                .map(|(k, v)| format!("{}: {}", k, truncate_json_fields(v, max_len)))
                .collect();
            format!("{{{}}}", entries.join(", "))
        }
        Value::Array(arr) => {
            let items: Vec<String> = arr
                .iter()
                .take(3)
                .map(|v| truncate_json_fields(v, max_len))
                .collect();
            if arr.len() > 3 {
                format!("[{}, ...+{}]", items.join(", "), arr.len() - 3)
            } else {
                format!("[{}]", items.join(", "))
            }
        }
        Value::String(s) => {
            if s.len() > max_len {
                format!("\"{}...\"", crate::truncate::truncate_str(s, max_len))
            } else {
                format!("\"{}\"", s)
            }
        }
        other => other.to_string(),
    }
}

// Global singleton
static COST_TRACKER: std::sync::OnceLock<CostTracker> = std::sync::OnceLock::new();

pub fn get_cost_tracker() -> &'static CostTracker {
    COST_TRACKER.get_or_init(CostTracker::new)
}

pub fn shutdown_cost_tracker() {
    if let Some(tracker) = COST_TRACKER.get() {
        tracker.flush();
    }
}
